### PR TITLE
Feature/7 restore with velero

### LINKED ISSUE #7: Implement manual restores with velero

### PR DESCRIPTION
**Pull request by jelemux**
Resolves #7

### CODE REVIEW

File: pkg/maintenance/looseCoupling.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+	watch, err := lcms.statefulSetClient.Watch(ctx, metav1.ListOptions{FieldSelector: "metadata.name=etcd"})
+	if err != nil {
+		return fmt.Errorf("failed to create watch for StatefulSet etcd: %w", err)
+	}
+
+	defer watch.Stop()
+
+	result := make(chan error, 1)
+	go func(ctx context.Context) {
```
Comment (cesmarvin) [inline pkg/maintenance/looseCoupling.go:95]: 🚫 **[checkstyle]** <contextcheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Non-inherited new context, use function like `context.WithXXX` instead

File: pkg/maintenance/looseCoupling.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		maintenanceModeSwitch: New(globalConfig),
+		statefulSetClient:     clientSet,
+		serviceInterface:      serviceInterface,
+	}
+}
+
+// ActivateMaintenanceMode activates the maintenance mode if the etcd exists and is ready.
+// This loose coupling enables us to perform restores on an empty cluster.
+func (lcms *looselyCoupledMaintenanceSwitch) ActivateMaintenanceMode(ctx context.Context, title string, text string) error {
+	if etcdReady, err := lcms.isEtcdReady(); err != nil {
```
Comment (cesmarvin) [inline pkg/maintenance/looseCoupling.go:35]: 🚫 **[checkstyle]** <contextcheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Function `isEtcdReady` should pass the context parameter

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,10 +5,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Added
+- [#7] Functionality to restore a backup to the namespace where the backup-operator is deployed
+    - Before the restore is applied, resources in this namespace which are irrelevant to the backup process are removed to provide a clean slate
```
Comment (nhinze23) [inline CHANGELOG.md:10]:
```suggestion
    - Before the restore is applied, resources in this namespace which are irrelevant to the backup process are removed to provide a clean state
```

**Review by jelemux — COMMENTED** [review]
LGTM!

File: main.go [hunk 4]
```diff
@@ -100,13 +91,17 @@ func configureManager(k8sManager controllerManager, operatorConfig *config.Opera
 }
 
 func getK8sManagerOptions(operatorConfig *config.OperatorConfig) ctrl.Options {
+	leaseDuration := time.Second * 60
```
Comment (nhinze23) [inline main.go:94]: Make both as a constant

File: pkg/api/v1/restore_types.go [hunk 5]
```diff
@@ -54,3 +61,7 @@ type RestoreList struct {
 func init() {
 	SchemeBuilder.Register(&Restore{}, &RestoreList{})
 }
+
```
Comment (nhinze23) [inline pkg/api/v1/restore_types.go:64]: Missing comment

File: pkg/backup/backupReconciler.go [hunk 6]
```diff
@@ -51,7 +46,7 @@ func (r *backupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 	logger := log.FromContext(ctx)
 
 	backup, err := r.clientSet.EcosystemV1Alpha1().Backups(r.namespace).Get(ctx, req.Name, metav1.GetOptions{})
-	if client.IgnoreNotFound(err) != nil {
```
Comment (nhinze23) [inline pkg/backup/backupReconciler.go:54]: Log on ignore not found

File: pkg/cleanup/cleanup.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		},
+	},
+}
+
+type defaultCleanupManager struct {
+	namespace       string
+	client          k8sClient
+	discoveryClient discoveryInterface
+}
+
```
Comment (nhinze23) [inline pkg/cleanup/cleanup.go:45]: Missing comment

File: pkg/cleanup/cleanup.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+func NewManager(namespace string, client k8sClient, discoveryClient discoveryInterface) Manager {
+	return &defaultCleanupManager{namespace: namespace, client: client, discoveryClient: discoveryClient}
+}
+
```
Comment (nhinze23) [inline pkg/cleanup/cleanup.go:49]: Same

File: pkg/cleanup/cleanup.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	return nil
+}
+
+func (c *defaultCleanupManager) deleteApiResourcesByLabelSelector(ctx context.Context, list *metav1.APIResourceList, selector labels.Selector) []error {
+	if len(list.APIResources) == 0 {
+		return nil
+	}
+	gv, err := schema.ParseGroupVersion(list.GroupVersion)
+	if err != nil {
+		return nil
```
Comment (nhinze23) [inline pkg/cleanup/cleanup.go:84]: Why we don't throw an error here? Maybe log in this situation.

File: pkg/cleanup/interfaces.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
+package cleanup
+
+import (
+	"context"
+
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
```
Comment (nhinze23) [inline pkg/cleanup/interfaces.go:9]: Comments

File: pkg/restore/requeueHandler.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
+package restore
```
Comment (nhinze23) [inline pkg/restore/requeueHandler.go:1]: The requeueHandler maybe could be generic

File: pkg/velero/restoreManager.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+				return fmt.Errorf("failed to complete velero restore [%s]: has status phase [%s]", changedRestore.Name, changedRestore.Status.Phase)
+			case velerov1.RestorePhaseCompleted:
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
```
Comment (nhinze23) [inline pkg/velero/restoreManager.go:111]: Comment

**Review by jelemux — COMMENTED** [review]
LGTM!

**Pull request merged**